### PR TITLE
fix: add prevent destroy and nullResource dependents for event-rules and SNS topics

### PR DESCRIPTION
**Pull request by sri-kirsh**
## Goal
add prevent destroy and nullResource dependents for event-rules and SNS topics

why?
- aws doesn't prevent sns topic and event-rule from getting deleted even if they are tied to a target

[INFRA-874](https://getpocket.atlassian.net/browse/INFRA-874)

[INFRA-874]: https://getpocket.atlassian.net/browse/INFRA-874?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[INFRA-874]: https://getpocket.atlassian.net/browse/INFRA-874?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: .aws/src/event-rules/user-api-events/userApiEventRules.ts [hunk 1]
```diff
@@ -4,27 +4,60 @@ import {
...
+      this.snsTopicDlq.name,
+      `${eventConfig.name}-Rule-dlq-alarm`,
+      4,
+      300,
+      10
+    );
+
+    //place-holder resource used to make sure we are not
+    //removing the event-rule or the SNS by mistake
+    //if the resources are removed, this would throw error
```
Comment (sri-kirsh) [inline .aws/src/event-rules/user-api-events/userApiEventRules.ts:55]:
```suggestion
    //if the resources are removed, this would throw compilation error
```

File: .aws/src/event-rules/user-api-events/userApiEventRules.ts [hunk 2]
```diff
@@ -4,27 +4,60 @@ import {
...
+      this,
+      pagerDuty,
+      this.snsTopicDlq.name,
+      `${eventConfig.name}-Rule-dlq-alarm`,
+      4,
+      300,
+      10
+    );
+
+    //place-holder resource used to make sure we are not
```
[the remark this answers is not in the document]
Comment (kschelonka) [reply .aws/src/event-rules/user-api-events/userApiEventRules.ts:53]: Ok thank you that makes sense to me, appreciate the comment

**Review by kschelonka — APPROVED** [review]

**Review by cmharlow — CHANGES_REQUESTED** [review]
I'm working on a fuller explanation of what we discussed in Slack vs what I see setup here (as they're different) to try to confirm the desired behavior. I'll try to have that as a fuller PR review in the next hour or so. But want to block merging this until I can confirm @sri-kirsh the expected behavior from this PR and why it differs from the Slack discussion (namely, lack of lifecycle on null-resources).

File: .aws/src/event-rules/account-delete-monitor/index.ts [hunk 3]
```diff
@@ -33,9 +40,23 @@ export class AccountDeleteMonitorEvents extends Resource {
 
     this.UserMergeTopic = new sns.SnsTopic(this, 'user-merge-topic', {
       name: `${config.prefix}-${admConfig.userMerge.name}-Topic`,
+      lifecycle: {
```
Comment (cmharlow) [inline .aws/src/event-rules/account-delete-monitor/index.ts:43]: Alright, I'm going to add a review comment here, but this largely applies to all 4 groups of resources updated in this PR ("groups" here being adm, premium-purchase, prospect-events, user-api-events).

my understanding of the issue is:
1. that AWS allows you to delete SNS Topics and eventRules even if there are targets attached to them; 
2. and there is no way for Terraform to apply lifecycle rules (e.g. `lifecycle.preventDestroy`) to the `aws_sns_topic` and `aws_cloudwatch_event_rule` Terraform resources (so we could prevent resource deletion at the Terraform level, since we could not prevent deletion at the AWS level).

However, I see that this PR (and the Terraform generated from it) _does_ add lifecycle rules (e.g. `lifecycle.preventDestroy`) to the `aws_sns_topic` and `aws_cloudwatch_event_rule` Terraform resources - and these rules seem to work (e.g. running plans that would delete these resources throws a Terraform error).

With that applied, I'm not sure we need to add a dependent, empty `null_resource`, which was only to cover part 2 of the issue above (which appears to be incorrect). If we want to follow the discussion in Slack, it would be these `null_resource` resources that had lifecycle rules applied to them (e.g. `lifecycle.preventDestroy`) as well as the dependency on the classes of resources that we thought couldn't accept Terraform lifecycle rules, so the `null_resource` would be a sort of "proxy" for the desired preventDelete behavior on those 2 resource types (`aws_sns_topic` and `aws_cloudwatch_event_rule`).

Below I walk through some examples of delete scenarios with the two resource types (`aws_sns_topic` and `aws_cloudwatch_event_rule`) in mind to confirm my understanding that the lifecycle rules are sufficient (though there is certainly no harm in leaving the `null_resources` around if you like, its just superfluous). 

### delete scenario 1: resource deletion through attr change that forces recreation

If I manually edit the generated Terraform json to change the `name` field on an `aws_sns_topic` - an attribute change that will cause a resource plan to involve a Terraform recreate (e.g. destroy & create), then run a terraform plan, Terraform errors out because of the `lifecycle.preventDestroy` attribute on this resource. Example output from a terraform plan after doing the manual name change on a `aws_sns_topic` is below:

```
╷
│ Error: Instance cannot be destroyed
│
│   on cdk.tf.json line 784, in resource.aws_sns_topic:
│  784:       "pocketeventbridge_admevents_usermergetopic_DEB7603D": {
│
│ Resource aws_sns_topic.pocketeventbridge_admevents_usermergetopic_DEB7603D has lifecycle.prevent_destroy set, but the plan calls for this resource to be destroyed. To avoid
│ this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan using the -target flag.
╵
```

Here's the same scenario but with a protected `aws_cloudwatch_event_rule`:

```
╷
│ Error: Instance cannot be destroyed
│
│   on cdk.tf.json line 405, in resource.aws_cloudwatch_event_rule:
│  405:       "pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C": {
│
│ Resource aws_cloudwatch_event_rule.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C has lifecycle.prevent_destroy set, but
│ the plan calls for this resource to be destroyed. To avoid this error and continue with the plan, either disable lifecycle.prevent_destroy or reduce the scope of the plan
│ using the -target flag.
```




### delete scenario 2: resource is explicitly deleted somehow

I hope this scenario would be easily caught / validated by the Typescript abstractions we have in place, but if not, i've manually edited the cdk.tf.json to remove a `aws_sns_topic` resource and tried to run a Terraform plan, but hit an error:

```
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 45, in data.aws_iam_policy_document.pocketeventbridge_admevents_PocketEventBridgeProdEventBridgeSNSPolicy_41C0E589.statement[0].resources:
│   45:               "${aws_sns_topic.pocketeventbridge_admevents_usermergetopic_DEB7603D.arn}"
│
│ A managed resource "aws_sns_topic" "pocketeventbridge_admevents_usermergetopic_DEB7603D" has not been declared in the root module.
╵
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 498, in resource.aws_cloudwatch_event_target.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_eventbridgetargetAccountDeleteMonitorProdUserMergeRuleTarget_3AA65A81:
│  498:         "arn": "${aws_sns_topic.pocketeventbridge_admevents_usermergetopic_DEB7603D.arn}",
│
│ A managed resource "aws_sns_topic" "pocketeventbridge_admevents_usermergetopic_DEB7603D" has not been declared in the root module.
╵
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 503, in resource.aws_cloudwatch_event_target.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_eventbridgetargetAccountDeleteMonitorProdUserMergeRuleTarget_3AA65A81.depends_on:
│  503:           "aws_sns_topic.pocketeventbridge_admevents_usermergetopic_DEB7603D",
│
│ A managed resource "aws_sns_topic" "pocketeventbridge_admevents_usermergetopic_DEB7603D" has not been declared in the root module.
╵
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 984, in resource.null_resource.pocketeventbridge_admevents_nullresource_582607DC.depends_on:
│  984:           "aws_sns_topic.pocketeventbridge_admevents_usermergetopic_DEB7603D"
│
│ A managed resource "aws_sns_topic" "pocketeventbridge_admevents_usermergetopic_DEB7603D" has not been declared in the root module.
```

You _could_ delete this topic if you also then deleted all references to this topic in those 3 other places - `data.aws_iam_policy_document`,  `resource.aws_cloudwatch_event_target`, & our new`resource.null_resource`.


And here is the same situation, but after fully deleting a protected `aws_cloudwatch_event_rule`:

```
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 490, in resource.aws_cloudwatch_event_target.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_eventbridgetargetAccountDeleteMonitorProdUserMergeRuleTarget_3AA65A81.depends_on:
│  490:           "aws_cloudwatch_event_rule.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C"
│
│ A managed resource "aws_cloudwatch_event_rule" "pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C" has not been declared in
│ the root module.
╵
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 493, in resource.aws_cloudwatch_event_target.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_eventbridgetargetAccountDeleteMonitorProdUserMergeRuleTarget_3AA65A81:
│  493:         "rule": "${aws_cloudwatch_event_rule.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C.name}",
│
│ A managed resource "aws_cloudwatch_event_rule" "pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C" has not been declared in
│ the root module.
╵
╷
│ Error: Reference to undeclared resource
│
│   on cdk.tf.json line 981, in resource.null_resource.pocketeventbridge_admevents_nullresource_582607DC.depends_on:
│  981:           "aws_cloudwatch_event_rule.pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C",
│
│ A managed resource "aws_cloudwatch_event_rule" "pocketeventbridge_admevents_PocketEventBridgeProdUserMergeEventBridgeRule_eventbridgerule_9306063C" has not been declared in
│ the root module.
```

You _could_ delete this eventRule if you also then deleted all references to this topic in those 2 other places - `resource.aws_cloudwatch_event_target`, & our new`resource.null_resource`.




### delete scenario 3: resource is implicitly deleted somehow

At least at the Terraform level in this setup, `aws_sns_topic` and `aws_cloudwatch_event_rule` aren't dependent on any "parent" resources which could be deleted and take their "children" resources out with them.



### conclusion

so the scenarios we're worried about protecting against:

1. some attribute change on one of the 2 resource types causes a recreate - and this is protected against via the lifecycle rules already applied to these resources.
2. someone accidentally deleting the code representations of one of these 2 resource types along with all references to the instances deleted - hopefully this would be caught by the code abstractions, but could have additional protection via code-isolated `null_resource` resources that have code-isolated dependencies on the types. This feels like a stretch, but we can do this if it is a concern.

**Review by cmharlow — COMMENTED** [review]
See my far too long comment (which I separated from this review) to try to grasp what is going on in the PR.

tl;dr, I won't block this, but I want to make sure I understand the addition of both lifecycle rules to the resources requiring protection _as well as_ null_resources with dependencies on those resources requiring protection.

**Review by cmharlow — APPROVED** [review]

**Comment by sri-kirsh** [comment]
had a discussed with Christina - we added null resource initially coz didnt realise preventDestroy was inside lifeCycle...but added lifecycle.preventDestroy after learning about it to sns/eventBridge

decided to leave nullResouce as it is helpful in giving a compilation error if sns/eventRule was removed before running a terraform plan...can remove it if its an additional overhead

not adding preventDestroy to nullResources coz it can get annoying later if we want to remove the null resource

**Review by cmharlow — APPROVED** [review]
